### PR TITLE
llext: add threads and kernel object test case

### PR DESCRIPTION
**Pull request by laurenmurphyx64**
Tests threads and kernel objects with userspace.

### CODE REVIEW

File: tests/subsys/llext/simple/src/threads_kernel_objects_ext.c [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * relocation support.
+ */
+
+#include <stdint.h>
+#include <zephyr/llext/symbol.h>
+#include <zephyr/kernel.h>
+
+#ifdef CONFIG_USERSPACE
+extern struct k_sem my_sem_userspace;
+extern struct k_thread my_thread_userspace;
```
Comment (pillo79) [inline tests/subsys/llext/simple/src/threads_kernel_objects_ext.c:18]: I would (re)use the local names even in the userspace case. With a few extra defines such as `MY_THREAD_PRIO`, `MY_THREAD_FLAGS` etc, most if not all of the other `#ifdefs` could go away :slightly_smiling_face: 
These `extern`s should also probably join other things in the shared .h for type safety.

File: tests/subsys/llext/simple/src/threads_kernel_objects_ext.c [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+#else
+static struct k_thread my_thread;
+static struct k_sem my_sem;
+#endif
+
+extern k_thread_stack_t my_thread_stack;
+#define STACK_SIZE 1024
```
Comment (pillo79) [inline tests/subsys/llext/simple/src/threads_kernel_objects_ext.c:25]: Move to .h

**Review by pillo79 — COMMENTED** [review]
Nice additional test, thank you! Some minor feedback on this.

File: tests/subsys/llext/simple/src/threads_kernel_objects_ext.c (hunk 1 above)
Comment (laurenmurphyx64) [reply tests/subsys/llext/simple/src/threads_kernel_objects_ext.c]: Great suggestion, it's pretty messy as-is. Have put them in a header.

File: tests/subsys/llext/simple/src/test_llext_simple.c [hunk 3]
```diff
@@ -50,7 +51,25 @@ int z_impl_ext_syscall_ok(int a)
 	return a + 1;
 }
 
+/* threads_kernel_objects test */
+#define STACK_SIZE 1024
+K_THREAD_STACK_DEFINE(my_thread_stack, STACK_SIZE);
+EXPORT_SYMBOL(my_thread_stack);
+
 #ifdef CONFIG_USERSPACE
+K_SEM_DEFINE(my_sem_userspace, 1, 1);
```
[the remark this answers is not in the document]
Comment (laurenmurphyx64) [reply tests/subsys/llext/simple/src/test_llext_simple.c]: I think I fixed it, let me know if I misunderstood

**Comment by laurenmurphyx64** [comment]
The compliance check on this one is destined to fail because checkpatch can't detect I actually did export the variable after I declared it:

```
ERROR   : Test Checkpatch failed: 
EXPORT_SYMBOL: EXPORT_SYMBOL(foo); should immediately follow its function/variable
File:tests/subsys/llext/simple/src/test_llext_simple.c
Line:59
```

Lines 58 and 59:
```
K_THREAD_STACK_DEFINE(my_thread_stack, MY_THREAD_STACK_SIZE);
EXPORT_SYMBOL(my_thread_stack);
```

@teburd Should we add exports to the checkpatch ignore, or just let this one pass, or am I missing something...?

**Comment by teburd** [comment]
we should tell checkpatch to ignore EXPORT_SYMBOL imho

**Comment by laurenmurphyx64** [comment]
> we should tell checkpatch to ignore EXPORT_SYMBOL imho

@teburd Added a commit to do that, others welcome to weigh in on it

**Review by teburd — DISMISSED** [review]

**Review by pillo79 — COMMENTED** [review]
Thanks, this looks way better - basically just one comment from me.
 
I was still confused by all the remaining `#ifdef`s though, so I played around with the sample a bit. The result are two extra commits in my repo [here](https://github.com/pillo79/zephyr/tree/threads_objects_llext_sample); please take a look and feel free to integrate any bits or pieces from there in this PR if you think they are useful :slightly_smiling_face:

File: tests/subsys/llext/simple/src/test_llext_simple.c [hunk 4]
```diff
@@ -7,9 +7,11 @@
 #include <zephyr/ztest.h>
 #include <zephyr/kernel.h>
 #include <zephyr/llext/llext.h>
+#include <zephyr/llext/symbol.h>
 #include <zephyr/llext/buf_loader.h>
 #include <zephyr/logging/log.h>
 #include <zephyr/sys/libc-hooks.h>
+#include "threads_kernel_objects.h"
```
[the remark this answers is not in the document]
Comment (laurenmurphyx64) [reply tests/subsys/llext/simple/src/test_llext_simple.c]: Thanks! I took your commits and squashed them into my original + changed the name of threads_kernel_objects_ext.h / moved it below syscalls_ext.h

**Review by teburd — APPROVED** [review]

**Review by pillo79 — APPROVED** [review]
Love it!

**Pull request merged**